Use NamedWritable to enable GeoBoundingBox serialisation

craigtaverner opened the pull request:
`BoundingBox` allows inter-node serialisation only within aggs results as components of aggregations that themselves use `NamedWritable`. However, if a `GeoBoundingBox` is added to search `hits`, serialisation fails as this type is not explicitly supported by `StreamOuput/StreamInput`. This was never noticed before because the only way to get a `BoundingBox` into the `hits` is through a painless script, something that appears to have not been used by customers, and is only tested internally using a single-node, single-shard yaml test, which does not touch this code path. We only found this issue due to recent effort to run all yaml tests in a serverless environment, which appears to always require serialisation from backend data nodes to a coordinator node.

Having said that, it is a surface API that could be used by users, and so should be fixed. While the missing type could be added explicitly, a much more flexible approach is to allow named classes to encapsulate their own serialization/deserialization logic. Here we use `NamedWritable` for that purpose.

Tasks:

- [x] Implement GenericNamedWriteable with version checking capabilities (based on VersionedNamedWriteable)
- [x] Manual testing with a multi-node cluster to verify this fixes #99089
- [x] Unit tests for version checking
- [x] Unit tests for serialization of GeoBoundingBox
- [x] Support CartesianBoundingBox by registering the named writable in SpatialPlugin

Fixes #99089

craigtaverner committed:
Use NamedWritable to enable GeoBoundingBox serialization

BoundingBox allows inter-node serialization only within aggs results as components of aggregations that themselves use NamedWritable.
However, if a GeoBoundingBox is added to search hits serialization fails as this type is not explictly supported by StreamOuput/StreamInput.
This was never noticed before because the only way to get a BoundingBox into the hits is through a painless script, something that appears to have not been used by customers, and is only tested internally using a single-node, single-shard yaml test, which does not touch this code path.
We only found this issue due to recent effort to run all yaml tests in a serverless environment, which appears to always require serialization from backend data nodes to a coordinator node.

Having said that, it is a surface API that could be used by users, and so should be fixed.
While the missing type could be added explicitly, a much more flexible approach is to allow named classes to encapsulate their own serialization/deserialization logic.
Here we use NamedWritable for that purpose.

Fixes #99089

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

craigtaverner committed:
Update docs/changelog/99163.yaml

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -789,6 +789,7 @@ public Object readGenericValue() throws IOException {
             case 27 -> readOffsetTime();
             case 28 -> readDuration();
             case 29 -> readPeriod();
+            case 30 -> readNamedWriteable(NamedWriteable.class);
```
This makes me a little uncomfortable, we should use a more precise class than just `NamedWriteable`. But I see why it's tricky here when we really don't know the context in which we're reading.

Could we introduce a marker type called `GenericNamedWriteable` and use that instead of `NamedWriteable`? That would help to protect against just accepting absolutely any `NamedWriteable` here.

Also I think this counts as a new transport version, we need to add a new version constant and throw an exception if the remote node version is too old rather than sending it something it won't understand.

craigtaverner replied:
Great advice. I'm creating GenericNamedWriteable to extend VersionedNamedWriteable, which also provides the version checking.

DaveCTurner replied:
I recommend not using `VersionNamedWriteable`, it is kind of a bad idea (it's designed to just silently drop stuff when sending to older versions, and that's almost certainly a mistake).

DaveCTurner replied:
Actually wait, I see, it could be useful here. But we do need to throw something (and ideally `assert false` too) if the sender is trying to send a value that the receiver won't understand.

craigtaverner committed:
Use GenericNamedWriteable to limit this to specific classes

So only classes that implement this specific interface will benefit from this additional capability.
Also we implement VersionedNamedWriteable to ensure that TransportVersion checking is done.

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -782,6 +782,11 @@ public final void writeOptionalInstant(@Nullable Instant instant) throws IOExcep
             o.writeInt(period.getMonths());
             o.writeInt(period.getDays());
+        }),
+        entry(GenericNamedWriteable.class, (o, v) -> {
+            o.writeByte((byte) 30);
+            GenericNamedWriteable w = (GenericNamedWriteable) v;
+            o.writeNamedWriteable(w);
         })
```
This is the extra validation I think we should have:

```suggestion
        entry(GenericNamedWriteable.class, (o, v) -> {
            final var genericNamedWriteable = (GenericNamedWriteable) v;
            final var minSupportedVersion = TransportVersion.max(
                TransportVersion.V_8_500_069,
                genericNamedWriteable.getMinimalSupportedVersion()
            );

            if (o.getTransportVersion().before(minSupportedVersion)) {
                final var message = Strings.format(
                    "[%s] requires minimal transport version [%s] and cannot be sent using transport version [%s]",
                    genericNamedWriteable,
                    minSupportedVersion,
                    o.getTransportVersion()
                );
                assert false : message;
                throw new IllegalStateException(message);
            }

            o.writeByte((byte) 30);
            o.writeNamedWriteable(genericNamedWriteable);
        })
```

craigtaverner replied:
That code looks good, but does it really do anything different to https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/io/stream/VersionCheckingStreamOutput.java#L64? That code should be called already, inside the call to `writeNamedWritable`, so the check already exists, as far as I can tell.

craigtaverner replied:
What I would like would be compatibility checks that verify that the exception is indeed thrown when we have mix-clusters. I wrote some tests like this before (in other work), but they were flaky (throwing the expected exception 99% of the time, not 100%, which was odd). I can try again.

DaveCTurner replied:
`VersionCheckingStreamOutput` doesn't consider it a bug to try and send an unknown writable, it just throws an IAE which maps onto a `400 Bad Request` and isn't logged. I think that's unduly lenient (although I see why checking this properly is tricky in a cross-cluster situation) and I'd rather we were stricter where possible, asserting that this doesn't happen and throwing an ISE if ever it does.

DaveCTurner replied:
> What I would like would be compatibility checks that verify that the exception is indeed thrown when we have mix-clusters.

I think a unit test is enough for this. We shouldn't even be trying to send unknown stuff over the wire ideally, hence why I think we should assert that we don't. Instead I'd rather that we looked at `org.elasticsearch.cluster.ClusterState#getMinTransportVersion` and avoided even creating the problematic object until all nodes can understand it.

craigtaverner replied:
I'm thinking, if we have a mixed cluster, and there is a query involving a new feature that requires sending something new over the wire, that only some of the cluster understands, the consequence is:

* If the sender is an old version, it will not even get to this point because the new feature is not understood causing an exception in the feature itself (eg. in my case the painless script will fail with a painless exception).
* If the sender is new, but the receiver is old, then the sender serialises type=30, and the receiver throws `IOException("Can't read unknown type [" + type + "]")` in `StreamInput.readGenericValue`.

So, as far as I can tell your suggestion is that instead of the receiver throwing this exception, we have a sender that detects the receiver is old and throws an exception before sending, which moves the error to a different node in the cluster. Is this a big improvement?

In fact there are three scenarios:

* If we do no checking of versions at all (ie. my initial code) we get the behaviour described above
* If we check with VersionedNamedWritable (ie. my current code) we get an exception in the sender if the VersionedNamedWritable version is later than the senders TransportVersion which only considers local node state, so does not care if the remote node has an older version
* We check the remote transport version also, and throw an exception there. Looking at your code, it does not do this but seems to consider local transport version only, so very similar to the current behaviour. Perhaps your last suggestion of using `ClusterState#getMinTransportVersion` is the way to do this? But again, the consequence is only to move the exception to a different node in the cluster.

craigtaverner replied:
I've done a little more digging and it looks like VersionedNamedWritable checks only occur for CCS, way too narrow a scope for our needs, so it is clear that I do need to write the check you have suggested. I'll go for that now.

DaveCTurner replied:
We should forbid the use of any features that the whole cluster does not understand (which is why I think we should be looking at `ClusterState#getMinTransportVersion`). At least, that's [the documented behaviour](https://www.elastic.co/guide/en/elastic-stack/8.9/upgrading-elasticsearch.html#rolling-upgrades). Then we should consider it a bug (and `assert false`) if ever we get into a position where we're trying to send a message that the remote won't understand.

> Looking at your code, it does not do this but seems to consider local transport version only

No, it's looking at `StreamOutput#getTransportVersion` which is a transport version compatible with both sender and receiver.

craigtaverner replied:
I've investigated further and figured out why I misunderstood this. I did not see the uses of `setTransportVersion`, which allows for the behaviour you describe.

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/geo/BoundingBox.java:
```diff
@@ -194,4 +195,9 @@ public int hashCode() {
     }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersion.V_8_500_069;
+    }
```
I think it'd be better to have this in the concrete subclasses. That way if we add another one in future we won't forget to set this to the correct value.

Also it (and the `getWriteableName` implementations) probably should to be `final` to make sure we don't accidentally add further subclasses without thinking about the consequences in this area..

craigtaverner replied:
Done

craigtaverner committed:
Code style fix

craigtaverner committed:
Explicit checks of version support for GenericNamedWritable

Co-authored-by: David Turner <david.turner@elastic.co>

craigtaverner commented on server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -782,6 +783,27 @@ public final void writeOptionalInstant(@Nullable Instant instant) throws IOExcep
+        }),
+        entry(GenericNamedWriteable.class, (o, v) -> {
+            final var genericNamedWriteable = (GenericNamedWriteable) v;
+            final var minSupportedVersion = TransportVersion.max(
```
This will always return the value of `genericNamedWriteable.getMinimalSupportedVersion()` since the concept of `GenericNamedWritable` was only introduced at this version.

DaveCTurner replied:
That's true, as long `getMinimalSupportedVersion()` is implemented correctly and not (say) copied from one of the implementations that returns `TransportVersion.MINIMUM_COMPATIBLE`. However BwC serialization is fairly subtle, and bugs in this area are just awful, so I'd like the extra check.

(We could also reasonably assert that `genericNamedWriteable.getMinimalSupportedVersion().onOrAfter(V_8_500_069)`)

craigtaverner replied:
I've added the assert now

craigtaverner commented on server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -782,6 +783,27 @@ public final void writeOptionalInstant(@Nullable Instant instant) throws IOExcep
+                    minSupportedVersion,
+                    o.getTransportVersion()
+                );
+                assert false : message;
```
These asserts seem to cause integration tests to fail in more opaque ways, with the test code receiving less information about the failure. Is this generally a good idea? Is the advantage that we fail faster, and so avoid being mislead by later errors?

DaveCTurner replied:
We definitely want to fail here in tests, and simply throwing an exception won't necessarily do that because exceptions can be caught. It's unfortunate that we have less context to share here, although I would expect these things to be fairly easy to reproduce.

craigtaverner replied:
At least here we assert with the same message. The test failure I debugged recently failed without the message, hiding the reason for failure, making it much harder to track initially.

craigtaverner committed:
Simplify GenericNamedWriteable checks

craigtaverner committed:
Move NamedWriteable methods into concrete classes

craigtaverner committed:
Added assert on GenericNamedWriteable min supported version

craigtaverner committed:
Unit tests for serialization of GenericNamedWriteable

craigtaverner committed:
Ensure serialization tests use the generic code path

The parent class AbstractNamedWriteableTestCase was not using the writeGenericValue and readGenericValue that this work changed, so we override that behaviour to ensure we're testing the new code.

We also add a new test method for negative testing of errors thrown when trying to serialize using an older TransportVersion.
We do not throw similar errors on reading, with the assumption that the reader can always read things it knows about, regardless of the age of the sender.
Taking account the TransportVersion in the reader only make sense if reading a known type where the format has changed.

craigtaverner committed:
Support CartesianBoundingBox in StreamOutput/Input

This is one reason why we went for a generic approach, allowing plugins to register additional GenericNamedWriteables that can be serialized/deserialized by the StreamInput/StreamOutput code without the server depending on xpack.

DaveCTurner commented on server/src/main/java/org/elasticsearch/TransportVersion.java:
```diff
@@ -182,6 +182,8 @@ private static TransportVersion registerTransportVersion(int id, String uniqueId
     public static final TransportVersion V_8_500_068 = registerTransportVersion(8_500_068, "2683c8b4-5372-4a6a-bb3a-d61aa679089a");
     public static final TransportVersion V_8_500_069 = registerTransportVersion(8_500_069, "5b804027-d8a0-421b-9970-1f53d766854b");
+    public static final TransportVersion V_8_500_070 = registerTransportVersion(8_500_070, "6BADC9CD-3C9D-4381-8BD9-B305CAA93F86");
+
```
```suggestion
```

nit: no extra whitespace here please

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -760,6 +762,24 @@ public final void writeOptionalInstant(@Nullable Instant instant) throws IOExcep
+            // Note that we do not rely on the checks in VersionCheckingStreamOutput because that only applies to CCS
+            final var genericNamedWriteable = (GenericNamedWriteable) v;
+            TransportVersion minSupportedVersion = genericNamedWriteable.getMinimalSupportedVersion();
+            assert minSupportedVersion.onOrAfter(V_8_500_070) : "[GenericNamedWriteable] requires [" + V_8_500_070 + "]";
```
Might be worth mentioning `minSupportedVersion` in the message too.

craigtaverner replied:
Since this was in an assert that would only happen in tests, and therefor tracked by a developer who will easily follow the code, I favoured brevity (and a non-wrapping line) over detail.

The next line that does the real check for production code is much more informative (for both the user and any developer tracking a failure).

DaveCTurner replied:
Makes sense, thanks.

DaveCTurner commented on server/src/test/java/org/elasticsearch/search/geo/GeoBoundsGenericWriteableTests.java:
```diff
@@ -0,0 +1,120 @@
+
+            // Test writing fails with older version
+            try {
+                copyInstance(testInstance, older, valid);
```
I think we should just verify that writing to an older stream fails, no need to make a `copyInstance` which tries to read it back again.

craigtaverner replied:
OK. After simplifying the test I replaced this with just the writing code.

DaveCTurner commented on server/src/test/java/org/elasticsearch/search/geo/GeoBoundsGenericWriteableTests.java:
```diff
@@ -0,0 +1,120 @@
+            }
+
+            // Test reading succeeds with older version (we do not cripple the reader)
+            deserializedInstance = copyInstance(testInstance, valid, older);
```
I think this doesn't really make sense, we are always sure that reading and writing always use the same transport version, and the message we're creating isn't actually something an older node would be able to create.

craigtaverner replied:
OK. I've simplified this failure test to only test the failure case. The first part with (valid,valid) was already tested in another test `testSerialization`, and the last part is unnecessary:

DaveCTurner reviewed (commented):
Looks great, just a few small comments.

craigtaverner committed:
Update server/src/main/java/org/elasticsearch/TransportVersion.java

Co-authored-by: David Turner <david.turner@elastic.co>

craigtaverner committed:
Code review updates - simplify negative test to only test failure case

craigtaverner committed:
Merge branch 'supportNamedWritableSerializationBoundingBox' of github.com:craigtaverner/elasticsearch into supportNamedWritableSerializationBoundingBox

DaveCTurner commented on server/src/test/java/org/elasticsearch/search/geo/GeoBoundsGenericWriteableTests.java:
```diff
@@ -0,0 +1,113 @@
+        return new GeoBoundingBox(new GeoPoint(top, left), new GeoPoint(bottom, right));
+    }
+
+    @Override
+    protected GenericNamedWriteable copyInstance(GenericNamedWriteable original, TransportVersion version) throws IOException {
+        return copyInstance(original, version, version);
+    }
+
+    /**
+     * Read and write with different serialization versions
+     */
+    protected GenericNamedWriteable copyInstance(
+        GenericNamedWriteable original,
+        TransportVersion writeVersion,
+        TransportVersion readVersion
+    ) throws IOException {
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            output.setTransportVersion(writeVersion);
+            output.writeGenericValue(original);
+            try (StreamInput in = new NamedWriteableAwareStreamInput(output.bytes().streamInput(), getNamedWriteableRegistry())) {
+                in.setTransportVersion(readVersion);
+                return readGenericValue(in);
+            }
+        }
+    }
+
+    private GenericNamedWriteable readGenericValue(StreamInput in) throws IOException {
+        Object obj = in.readGenericValue();
+        if (obj instanceof GenericNamedWriteable result) {
+            return result;
+        }
+        throw new IllegalStateException("Read result of wrong type: " + obj.getClass().getSimpleName());
+    }
+
```
This is no longer needed, we can just use the `copyInstance` implementation in the subclass.

```suggestion
```

(also needs a spotless run to clean up imports)

DaveCTurner replied:
acked, I think I can see a way to make this a little less special, but let's merge this and I'll follow up with a suggestion.

DaveCTurner reviewed (approved):
LGTM (left a comment about some now-unnecessary test code tho)

craigtaverner committed:
Code review - remove unused method after simplifying test

DaveCTurner committed:
Simplify GeoBoundsGenericWriteableTests
